json_tricks/encoders.py

```diff
@@ -413,11 +415,12 @@ def numpy_encode(obj, primitives=False, properties=None):
 			if use_compact and store_endianness != 'suppress':
 				dct['endian'] = store_endianness or sys.byteorder
 			return dct
-	elif isinstance(obj, generic):
-		if NumpyEncoder.SHOW_SCALAR_WARNING:
-			NumpyEncoder.SHOW_SCALAR_WARNING = False
-			warnings.warn('json-tricks: numpy scalar serialization is experimental and may work differently in future versions')
-		return obj.item()
+	elif isinstance(obj, scalar_types):
+		dct = hashodict((
+			('__numpy_scalar__', obj.item()),
+			('dtype', str(obj.dtype)),
```

Comment:
This new approach seems great and indeed in an ideal world it would work perfectly.

...

However, it was found in issue #18 that Python sees some numpy scalars are primitives, and refuses to call encoders for them (presumably for performance).

Which ones depends on the Python version, for extra confusion, although I guess Python 2 is less important now.

In any case, this function in Python 3 works for a lot of types, but not for float64, which is an important one. There are two concerns with this:

* While keeping the numpy scalar type is better in general, doing it half the time seems like it adds more confusion than it's worth.
* We've been doing it this way, it's be a (slightly) breaking change to start doing it differently.

I think we'll need to think a bit more about this, maybe make it opt-in, or skip the scalars and do 0dimensional arrays only, if that's possible.

Reply:
omg, i feel so ashamed to not have tested float64.... 

I literally tested uint8 through uint64, int8 through int64, datetime64s, and float32.... but not float64....

for reference, this is the current output.
```python
{                                                                                   
    "uint32": {                                                                     
        "__numpy_scalar__": 1,                                                      
        "dtype": "uint32"                                                           
    },                                                                              
    "int32": {                                                                      
        "__numpy_scalar__": 1,                                                      
        "dtype": "int32"                                                            
    },                                                                              
    "float32": {                                                                    
        "__numpy_scalar__": 1.0,                                                    
        "dtype": "float32"                                                          
    },                                                                                                                                                                                                                
    "float64": 1.0,                                                                 
    "datetime64[ns]": {                                                         
        "__numpy_scalar__": 1704235669639528000,                                    
        "dtype": "datetime64[ns]"                                                   
    },                                                                              
    "datetime64[us]": {                                                             
        "__numpy_scalar__": {                                                       
            "__datetime__": null,                                                   
            "year": 2024,                                                           
            "month": 1,                                                             
            "day": 2,                                                               
            "hour": 22,                                                             
            "minute": 47,                                                           
            "second": 49,                                                           
            "microsecond": 639546                                                   
        },                                                                          
        "dtype": "datetime64[us]"                                                   
    }                                                                               
}
```